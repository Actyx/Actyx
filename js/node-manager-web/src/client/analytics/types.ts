import { FatalError } from "../../common/ipc";

export interface AnalyticsActions {
  viewedScreen: (screenName: string) => void;
  queriedEvents: (query: string) => void;
  startedApp: () => void;
  addedNode: () => void;
  removedNode: () => void;
  removedAllNodes: () => void;
  signedAppManifest: () => void;
  setSettings: () => void;
  shutdownNode: () => void;
  generatedSwarmKey: () => void;
  createdUserKeyPair: () => void;
  gotFatalError: (error: FatalError) => void;
  gotError: (errorMsg: string) => void;
}

export type AnalyticsEvent =
  | { readonly key: "ViewedScreen"; name: string }
  | { readonly key: "QueriedEvents"; query: string }
  | { readonly key: "StartedApp" }
  | { readonly key: "AddedNode" }
  | { readonly key: "RemovedNode" }
  | { readonly key: "RemovedAllNodes" }
  | { readonly key: "SignedAppManifest" }
  | { readonly key: "CreatedUserKeyPair"; wasDefault: boolean }
  | { readonly key: "GotFatalError"; shortMessage: string; details?: string }
  | { readonly key: "GotError"; shortMessage: string }
  | { readonly key: "SetSettings" }
  | { readonly key: "ShutdownNode" }
  | { readonly key: "GeneratedSwarmKey" };
