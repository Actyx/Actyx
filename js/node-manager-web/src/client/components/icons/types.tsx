import React from "react"

interface Props {
  width?: number
  height?: number
  className?: string
}

export type IconType = React.FC<Props>
