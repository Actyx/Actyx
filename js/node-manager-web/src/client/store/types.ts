import { StoreData as Data } from "../../common/types";

export const enum StoreStateKey {
  Loaded = "Loaded",
  LoadingOrSaving = "LoadingOrSaving",
  Initial = "Initial",
}

interface LoadedActions {
  updateAndReload: (new_: Data) => void;
  reload: () => void;
}

interface Loaded {
  key: StoreStateKey.Loaded;
  data: Data;
  actions: LoadedActions;
}
interface LoadingOrSaving {
  key: StoreStateKey.LoadingOrSaving;
}

interface Initial {
  key: StoreStateKey.Initial;
}

// Any error will actually be fatal for simplicity
export type StoreState = Initial | Loaded | LoadingOrSaving;

export const enum StoreActionKey {
  HasLoaded = "HasLoaded",
  LoadOrSave = " LoadOrSave",
}

interface HasLoaded {
  key: StoreActionKey.HasLoaded;
  data: Data;
  actions: LoadedActions;
}

interface LoadOrSave {
  key: StoreActionKey.LoadOrSave;
}

export type StoreAction = HasLoaded | LoadOrSave;

export type Dispatch = (action: StoreAction) => void;
