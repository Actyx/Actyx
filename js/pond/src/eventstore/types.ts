/*
 * Actyx Pond: A TypeScript framework for writing distributed apps
 * deployed on peer-to-peer networks, without any servers.
 * 
 * Copyright (C) 2020 Actyx AG
 */
import { right } from 'fp-ts/lib/Either'
import { Ord, ordNumber, ordString } from 'fp-ts/lib/Ord'
import { Ordering } from 'fp-ts/lib/Ordering'
import * as t from 'io-ts'
import { FishName, isString, Lamport, Psn, Semantics, SourceId, Timestamp } from '../types'
import { OffsetMapIO } from './offsetMap'
import { createEnumType, EnumType } from './utils'

export { OffsetMap, OffsetMapBuilder } from './offsetMap'

/**
 * Basically adds -infinity and +infinity to a PSN
 */
const PsnOrLimit = t.union([Psn.FromNumber, t.literal('min'), t.literal('max')])
export type PsnOrLimit = t.TypeOf<typeof PsnOrLimit>

/**
 * A psn map with a default value, so it is a tabulated total function from source to PsnOrLimit
 */
export const OffsetMapWithDefault = t.readonly(
  t.type({
    psns: t.readonly(OffsetMapIO),
    default: PsnOrLimit,
  }),
)
export type OffsetMapWithDefault = t.TypeOf<typeof OffsetMapWithDefault>

const stringRA = t.readonlyArray(t.string)

type Tags = ReadonlyArray<string>
type TagsOnWire = ReadonlyArray<string> | undefined
const Tags = new t.Type<Tags, TagsOnWire>(
  'TagsSetFromArray',
  (x): x is Tags => x instanceof Array && x.every(isString),
  // Rust side for now expresses empty tag arrays as omitting the field
  (x, c) => (x === undefined ? right([]) : stringRA.validate(x, c)),
  // Sending empty arrays is fine, though
  x => x,
)

export const EventIO = t.type({
  psn: Psn.FromNumber,
  semantics: Semantics.FromString,
  sourceId: SourceId.FromString,
  name: FishName.FromString,
  timestamp: Timestamp.FromNumber,
  lamport: Lamport.FromNumber,
  tags: Tags,
  payload: t.unknown,
})
export type Event = t.TypeOf<typeof EventIO>
const compareEvents = (a: Event, b: Event): Ordering => {
  const lamportOrder = ordNumber.compare(a.lamport, b.lamport)
  if (lamportOrder !== 0) {
    return lamportOrder
  }
  const sourceOrder = ordString.compare(a.sourceId, b.sourceId)
  if (sourceOrder !== 0) {
    return sourceOrder
  }
  return ordNumber.compare(a.psn, b.psn)
}

const eventsEqual = (a: Event, b: Event): boolean =>
  // compare numerical fields first since it should be faster
  a.lamport === b.lamport && a.psn === b.psn && a.sourceId === b.sourceId
/**
 * Order for events
 *
 * Order is [lamport, sourceId]
 * Events are considered equal when lamport, sourceId, psn are equal without considering
 * the content of the payload. Having two events that have these fields equal yet a different
 * payload would be a grave bug in our system.
 */
const ordEvent: Ord<Event> = {
  equals: eventsEqual,
  compare: compareEvents,
}
export const Event = {
  ord: ordEvent,
}

/**
 * A number of events, not necessarily from the same source
 */
export const Events = t.readonlyArray(EventIO)
export type Events = t.TypeOf<typeof Events>

/**
 * A number of generated events, that are going to be written to the store.
 */
export const UnstoredEvent = t.readonly(
  t.type({
    /**
     * the sequence nr of the first element in this chunk
     */
    semantics: Semantics.FromString,
    name: FishName.FromString,
    timestamp: Timestamp.FromNumber,
    tags: Tags,
    payload: t.unknown,
  }),
)

export type UnstoredEvent = t.TypeOf<typeof UnstoredEvent>
export const UnstoredEvents = t.readonlyArray(UnstoredEvent)
export type UnstoredEvents = t.TypeOf<typeof UnstoredEvents>

/**
 * Sort order for perstisted events
 */
export enum PersistedEventsSortOrders {
  EventKey = 'eventKey',
  ReverseEventKey = 'reverseEventKey',
  Unsorted = 'unsorted',
}
export const PersistedEventsSortOrder: EnumType<PersistedEventsSortOrders> = createEnumType<
  PersistedEventsSortOrders
>(PersistedEventsSortOrders, 'PersistedEventsSortOrders')
export type PersistedEventsSortOrder = t.TypeOf<typeof PersistedEventsSortOrder>

/**
 * Sort order for events
 */
export enum AllEventsSortOrders {
  EventKey = 'eventKey',
  Unsorted = 'unsorted',
}
export const AllEventsSortOrder: EnumType<AllEventsSortOrders> = createEnumType<
  AllEventsSortOrders
>(AllEventsSortOrders, 'AllEventsSortOrders')
export type AllEventsSortOrder = t.TypeOf<typeof AllEventsSortOrder>

/**
 * Connectivity status type.
 * @public
 */
export enum ConnectivityStatusType {
  FullyConnected = 'FullyConnected',
  PartiallyConnected = 'PartiallyConnected',
  NotConnected = 'NotConnected',
}

const FullyConnected = t.readonly(
  t.type({
    status: t.literal(ConnectivityStatusType.FullyConnected),
    inCurrentStatusForMs: t.number,
  }),
)

const PartiallyConnected = t.readonly(
  t.type({
    status: t.literal(ConnectivityStatusType.PartiallyConnected),
    inCurrentStatusForMs: t.number,
    specialsDisconnected: t.readonlyArray(t.string),
    swarmConnectivityLevel: t.number, // Percent*100, e.g. 50% would be 50, not 0.5
    eventsToRead: t.number,
    eventsToSend: t.number,
  }),
)

const NotConnected = t.readonly(
  t.type({
    status: t.literal(ConnectivityStatusType.NotConnected),
    inCurrentStatusForMs: t.number,
    eventsToRead: t.number,
    eventsToSend: t.number,
  }),
)

/**
 * The IO-TS type parser for ConnectivityStatus.
 * @public
 */
export const ConnectivityStatus = t.union([FullyConnected, PartiallyConnected, NotConnected])

/**
 * Current connectivity of the underlying ActyxOS node.
 * @public
 */
export type ConnectivityStatus = t.TypeOf<typeof ConnectivityStatus>

/* Other things */

/** Hook to run on store connection being closed. @public */
export type StoreConnectionClosedHook = () => void

/** Configuration for the WebSocket store connection. @public */
export type WsStoreConfig = Readonly<{
  /** url of the destination */
  url: string
  /** protocol of the destination */
  protocol?: string
  /** Hook, when the connection to the store is closed */
  onStoreConnectionClosed?: StoreConnectionClosedHook
  /** retry interval to establish the connection */
  reconnectTimeout?: number

  // todo timeouts?, heartbeats? etc.
}>
