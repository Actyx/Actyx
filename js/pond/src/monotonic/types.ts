import { Option } from 'fp-ts/lib/Option'
import { Events } from '../eventstore/types'
import { EventKey, LocalSnapshot, StateWithProvenance, Timestamp } from '../types'

// A local snapshot where the state has already been serialised
export type SerializedStateSnap = LocalSnapshot<string>

// A local snapshot pending persistence
export type PendingSnapshot = Readonly<{
  snap: SerializedStateSnap
  tag: string
  timestamp: Timestamp
}>

// A Reducer for the Events-Or-State message stream that
// - is unaware of serialization
// - can apply event chunks within bounds
export type SimpleReducer<S> = {
  // Apply the events between fromIdx and toIdxInclusive to the state
  appendEvents: (events: Events, fromIdx: number, toIdxInclusive: number) => LocalSnapshot<S>

  // Directly set the state
  setState: (state: LocalSnapshot<S>) => void
}

// A Reducer for the Events-Or-State message stream
// which will cache intermediate States.
// (This type would extend `Reducer`, if JS had 0-copy slices.)
export type CachingReducer<S> = {
  // Apply the given events to the state and return the new latest state.
  // Internally, this will queue up snapshots and persist them async (persistence can be awaited via awaitPendingPersistence)
  appendEvents: (events: Events) => StateWithProvenance<S>

  // Just directly set a certain state.
  // Will invalidate all later cached states.
  setState: (state: SerializedStateSnap) => void

  // Await persistence for all states (local snapshots)
  awaitPendingPersistence: () => Promise<void>

  // Get the latest state still valid according to the given arguments.
  // This method by itself does not invalidate anything inside the cache, it just filters.
  latestKnownValidState: (
    invalidateStatesAfter: EventKey,
    highestTrigger: EventKey,
  ) => Option<SerializedStateSnap>
}
