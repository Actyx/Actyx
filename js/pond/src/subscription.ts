/*
 * Actyx Pond: A TypeScript framework for writing distributed apps
 * deployed on peer-to-peer networks, without any servers.
 * 
 * Copyright (C) 2020 Actyx AG
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
import * as t from 'io-ts'
import { always } from 'ramda'
import { Event } from './eventstore/types'
import { Envelope, FishName, Semantics, SourceId } from './types'

const subscriptionToString = (subscription: Subscription): string => {
  if (subscription.name === '' && subscription.sourceId !== '') {
    throw new Error()
  }
  // TODO: escaping / and ,
  const parts = [subscription.semantics, subscription.name, subscription.sourceId].filter(
    x => x !== '',
  )
  return parts.join('/')
}

const parseSubscription = (text: string): Subscription => {
  const parts = text.trim().split('/')
  if (parts.length < 1 || parts.length > 3) {
    throw new Error()
  }
  const [semantics, name, sourceId] = parts
  return {
    semantics: Semantics.of(semantics),
    name: FishName.of(name),
    sourceId: SourceId.of(sourceId),
  }
}

export const SubscriptionIO = t.readonly(
  t.type({
    semantics: Semantics.FromString,
    // Wildcard is ""
    name: FishName.FromString,
    // Wildcard is ""
    sourceId: SourceId.FromString,
  }),
)
export const Subscription = {
  parse: parseSubscription,
  toString: subscriptionToString,
}
export type Subscription = t.TypeOf<typeof SubscriptionIO>

export const TagSubscription = t.readonly(
  t.type({
    tags: t.readonlyArray(t.string),
    local: t.boolean,
  }),
)

/**
 * A set of subscriptions
 */
export const SubscriptionSetIO = t.taggedUnion(
  // FIXME: why doesn't this work with t.readonly??
  'type',
  [
    t.type({ type: t.literal('empty') }),
    t.type({ type: t.literal('all') }),
    t.type({ type: t.literal('or'), subscriptions: t.readonlyArray(SubscriptionIO) }),
    t.type({ type: t.literal('tags'), subscriptions: t.readonlyArray(TagSubscription) }),
    // To be added in the future:
    // t.type({ type: t.literal('tag_expr'), subscriptions: t.string }),
  ],
)
export type SubscriptionSet = t.TypeOf<typeof SubscriptionSetIO>

export type SubscriptionSetCompanion = {
  empty: SubscriptionSet
  all: SubscriptionSet
  or: (s: ReadonlyArray<Subscription>) => SubscriptionSet
  parse: (text: string, separator?: string) => SubscriptionSet
}

const alwaysFalse = always(false)
const alwaysTrue = always(true)

const mkEnvelopePredicate: (s: Subscription) => (e: Envelope<any>) => boolean = s => {
  if (s.name !== '' && s.sourceId !== '') {
    // match everything
    return e =>
      e.source.semantics === s.semantics &&
      e.source.name === s.name &&
      e.source.sourceId === s.sourceId
  }
  if (s.sourceId !== '' && s.name === '') {
    // only name is ''
    return e => e.source.semantics === s.semantics && e.source.sourceId === s.sourceId
  }
  if (s.name !== '') {
    // match semantics and name
    return e => e.source.semantics === s.semantics && e.source.name === s.name
  }
  if (s.sourceId === '') {
    // match just semantics
    return e => e.source.semantics === s.semantics
  }
  throw new Error('Given combination of semantics / name / sourceId is not supported!')
}

const mkEventPredicate: (s: Subscription) => (e: Event) => boolean = s => {
  const envPredicate = mkEnvelopePredicate(s)
  // Just use Event in Envelope form.
  return e =>
    envPredicate({
      ...e,
      source: e,
    })
}

const mkSourceIdPredicate: (s: Subscription) => (sId: SourceId) => boolean = s => sId =>
  s.sourceId === '' ? true : s.sourceId === sId

const parseSubscriptions = (text: string, separator: string = '|'): SubscriptionSet => {
  const text1 = text.trim()
  if (text1 === 'all') {
    return SubscriptionSet.all
  } else if (text1 === 'empty') {
    return SubscriptionSet.empty
  } else {
    return SubscriptionSet.or(text.split(separator).map(Subscription.parse))
  }
}

export const SubscriptionSet: SubscriptionSetCompanion = {
  empty: { type: 'empty' },
  all: { type: 'all' },
  or: (subscriptions: ReadonlyArray<Subscription>) => ({ type: 'or', subscriptions }),
  parse: parseSubscriptions,
}

export const subscriptionsToPredicate: <T>(
  ss: SubscriptionSet,
  mkPredicate: (s: Subscription) => (t: T) => boolean,
) => (t: T) => boolean = (subscriptionSet, mkPredicate) => {
  switch (subscriptionSet.type) {
    case 'tags':
      throw new Error('tags case should not be passed to this function')
    case 'empty':
      return alwaysFalse
    case 'all':
      return alwaysTrue
    case 'or': {
      const { subscriptions } = subscriptionSet
      if (subscriptions.length === 0) {
        return alwaysFalse
      }
      if (subscriptions.length === 1) {
        return mkPredicate(subscriptionSet.subscriptions[0])
      }
      const predicates = subscriptions.map(mkPredicate)
      return ev => {
        /**
         * This used to be predicates.some(p => p(t)), but we want to avoid a closure
         * being created on each invocation of the predicate. It seems that v8 is not
         * smart enough to inline predicates.some()...
         */

        for (let i = 0; i < predicates.length; i++) {
          if (predicates[i](ev)) {
            return true
          }
        }
        return false
      }
    }
  }
}
export const subscriptionsToEnvelopePredicate = (ss: SubscriptionSet) => {
  if (ss.type === 'tags') {
    throw new Error('no tag support for envelopes!')
  }

  return subscriptionsToPredicate(ss, mkEnvelopePredicate)
}

export const subscriptionsToEventPredicate = (ss: SubscriptionSet) => {
  if (ss.type === 'tags') {
    // we can ignore the 'local' flag since it will never exclude our local events
    // and this method is used solely to decide whether locally emitted events are relevant
    return (event: Event) =>
      ss.subscriptions.some(tagIntersection =>
        tagIntersection.tags.every(tag => event.tags.includes(tag)),
      )
  }

  return subscriptionsToPredicate(ss, mkEventPredicate)
}

export const subscriptionsToSourceIdPredicate = (ss: SubscriptionSet) => {
  if (ss.type === 'tags') {
    throw new Error('tag-based subs cannot discern source id like this!')
  }

  return subscriptionsToPredicate(ss, mkSourceIdPredicate)
}
