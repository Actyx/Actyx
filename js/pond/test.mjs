import { FishId, Pond, Tags } from './lib/index.js'

const run = async () => {
  // const pond = await Pond.of({ appId: 'com.example.x', displayName: 'test', version: '1' }, {actyxHost:'what.ever'}, {})
  const pond = await Pond.default({ appId: 'com.example.x', displayName: 'test', version: '1' })
  console.log('got pond', pond)
  pond.observe(
    {
      fishId: FishId.of('test', 'test', 1),
      where: Tags(),
      initialState: 0,
      onEvent: (state, _event, meta) => {
        if (state % 1 == 0) console.log('onEvent', state, meta.timestampAsDate(), meta.tags)
        return state + 1
      },
      isReset: (_e, meta) => {
        const age = Math.floor(Date.now() - meta.timestampMicros / 1000)
        const reset = age > 1_000_000_000
        console.log('isReset', meta.timestampAsDate(), age, reset)
        return reset
      },
    },
    (state) => console.log('state', state),
  )
}
run()
