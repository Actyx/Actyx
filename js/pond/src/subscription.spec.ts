/* eslint-disable @typescript-eslint/no-explicit-any */
import { Subscription, SubscriptionSet, subscriptionsToEnvelopePredicate } from './subscription'
import { timerFishType } from './timerFish.support.test'
import { Envelope, FishName, Lamport, Semantics, Source, SourceId, Timestamp } from './types'

const sourceA: Source = {
  semantics: Semantics.of('a'),
  name: FishName.of(''),
  sourceId: SourceId.of('id'),
}
const sourceB: Source = {
  semantics: Semantics.of('b'),
  name: FishName.of(''),
  sourceId: SourceId.of('id'),
}
const evA: Envelope<any> = {
  source: sourceA,
  timestamp: Timestamp.of(1234),
  lamport: Lamport.of(1234),
  payload: {},
}
const evB: Envelope<any> = {
  source: sourceB,
  timestamp: Timestamp.of(1234),
  lamport: Lamport.of(1234),
  payload: {},
}

describe('subscriptionsToEnvelopePredicate', () => {
  it('must allow wildcards in all but semantics', () => {
    const sub = Subscription.of(sourceA.semantics)
    const pred = subscriptionsToEnvelopePredicate(SubscriptionSet.or([sub]))
    expect(pred(evA)).toBeTruthy()
    expect(pred(evB)).toBeFalsy()
  })
  it('must allow wildcards in name but fixed source id', () => {
    const sid = SourceId.of('id')
    const sub = Subscription.of(timerFishType, undefined, sid)
    expect(subscriptionsToEnvelopePredicate(SubscriptionSet.or([sub]))(evA)).toBeFalsy()
    const sub2 = Subscription.of(sourceA.semantics, undefined, sid)
    expect(subscriptionsToEnvelopePredicate(SubscriptionSet.or([sub2]))(evA)).toBeTruthy()
    expect(subscriptionsToEnvelopePredicate(SubscriptionSet.or([sub2]))(evB)).toBeFalsy()
  })
  it('must work for empty', () => {
    const pred = subscriptionsToEnvelopePredicate(SubscriptionSet.empty)
    expect(pred(evA)).toBeFalsy()
  })
  it('must work for all', () => {
    const pred = subscriptionsToEnvelopePredicate(SubscriptionSet.all)
    expect(pred(evA)).toBeTruthy()
  })
  it('must work for or with no subscription', () => {
    const pred = subscriptionsToEnvelopePredicate(SubscriptionSet.or([]))
    expect(pred(evA)).toBeFalsy()
  })
  it('must work for or with a single subscription', () => {
    const pred = subscriptionsToEnvelopePredicate(SubscriptionSet.or([sourceA]))
    expect(pred(evA)).toBeTruthy()
  })
  it('must work for or with multiple subscriptions', () => {
    const pred = subscriptionsToEnvelopePredicate(SubscriptionSet.or([sourceA, sourceB]))
    expect(pred(evA)).toBeTruthy()
    expect(pred(evB)).toBeTruthy()
  })
})

describe('SubscriptionSet.parse', () => {
  it('should parse subscription sets', () => {
    expect(SubscriptionSet.parse('all')).toEqual(SubscriptionSet.all)
    expect(SubscriptionSet.parse('empty')).toEqual(SubscriptionSet.empty)
    expect(SubscriptionSet.parse('a/b/c|d/e/f')).toMatchSnapshot()
    expect(SubscriptionSet.parse('a|b')).toMatchSnapshot()
  })
})
