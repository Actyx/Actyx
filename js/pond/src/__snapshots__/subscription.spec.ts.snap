// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionSet.parse should parse subscription sets 1`] = `
Object {
  "subscriptions": Array [
    Object {
      "name": "b",
      "semantics": "a",
      "sourceId": "c",
    },
    Object {
      "name": "e",
      "semantics": "d",
      "sourceId": "f",
    },
  ],
  "type": "or",
}
`;

exports[`SubscriptionSet.parse should parse subscription sets 2`] = `
Object {
  "subscriptions": Array [
    Object {
      "name": undefined,
      "semantics": "a",
      "sourceId": undefined,
    },
    Object {
      "name": undefined,
      "semantics": "b",
      "sourceId": undefined,
    },
  ],
  "type": "or",
}
`;
